src/services/codefixes/inferFromUsage.ts

```diff
@@ -0,0 +1,590 @@
+/* @internal */
+namespace ts.codefix {
+    registerCodeFix({
+        errorCodes: [
+            // Variable declarations
+            Diagnostics.Variable_0_implicitly_has_type_1_in_some_locations_where_its_type_cannot_be_determined.code,
+            Diagnostics.Variable_0_implicitly_has_an_1_type.code,
+
+            // Parameter declarations
+            Diagnostics.Parameter_0_implicitly_has_an_1_type.code,
+            Diagnostics.Rest_parameter_0_implicitly_has_an_any_type.code,
+
+            // Get Accessor declarations
+            Diagnostics.Property_0_implicitly_has_type_any_because_its_get_accessor_lacks_a_return_type_annotation.code,
+            Diagnostics._0_which_lacks_return_type_annotation_implicitly_has_an_1_return_type.code,
+
+            // Set Accessor declarations
+            Diagnostics.Property_0_implicitly_has_type_any_because_its_set_accessor_lacks_a_parameter_type_annotation.code,
+
+            // Property declarations
+            Diagnostics.Member_0_implicitly_has_an_1_type.code,
+        ],
+        getCodeActions: getActionsForAddExplicitTypeAnnotation
+    });
+
+    function getActionsForAddExplicitTypeAnnotation({ sourceFile, program, span: { start }, errorCode, cancellationToken }: CodeFixContext): CodeAction[] | undefined {
+        const token = getTokenAtPosition(sourceFile, start);
+
+        switch (token.kind) {
+            case SyntaxKind.Identifier:
+            case SyntaxKind.DotDotDotToken:
+            case SyntaxKind.PublicKeyword:
+            case SyntaxKind.PrivateKeyword:
+            case SyntaxKind.ProtectedKeyword:
+            case SyntaxKind.ReadonlyKeyword:
+                // Allowed
+                break;
+            default:
+                return undefined;
+        }
+
+        const containingFunction = getContainingFunction(token);
+        const checker = program.getTypeChecker();
+
+        switch (errorCode) {
+            // Variable and Property declarations
+            case Diagnostics.Member_0_implicitly_has_an_1_type.code:
+            case Diagnostics.Variable_0_implicitly_has_type_1_in_some_locations_where_its_type_cannot_be_determined.code:
+                return getCodeActionForVariableDeclaration(<PropertyDeclaration | PropertySignature | VariableDeclaration>token.parent);
+            case Diagnostics.Variable_0_implicitly_has_an_1_type.code:
+                return getCodeActionForVariableUsage(<Identifier>token);

```

Comment:
I feel like there should be assertions when you cast based on the error.

Reply:
We asserted a few lines earlier that it has to be an identifier, DotDotDotToken, PublicKeyword, ProtectedKeyword, or ReadonlyKeyword. since this is a variable declaration error, we know it can not be anything that is not an identifier (i.e. binding pattern).